### PR TITLE
Fix `ALTER DELETE` queries with materialized `_block_number` column

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed `ALTER DELETE` queries with materialized `_block_number` column (if setting `enable_block_number_column` is enabled).

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit b66330828ded68e6dde33dee9330b1186493d30a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72261/b66330828ded68e6dde33dee9330b1186493d30a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71085/15fa138776d420a2154dbe888f44dc1ed7fadd63/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/MutationsInterpreter.cpp [hunk 1]
```diff
@@ -414,10 +416,21 @@ MutationsInterpreter::MutationsInterpreter(
     ContextPtr context_,
     Settings settings_)
     : MutationsInterpreter(
-        Source(storage_, std::move(source_part_), std::move(alter_conversions_)),
+        Source(storage_, source_part_, std::move(alter_conversions_)),
         std::move(metadata_snapshot_), std::move(commands_),
         std::move(available_columns_), std::move(context_), std::move(settings_))
 {
+    const auto & part_columns = source_part_->getColumnsDescription();
```
Comment (SmitaRKulkarni) [inline src/Interpreters/MutationsInterpreter.cpp:423]: perhaps we should move the changes to this constructor https://github.com/ClickHouse/ClickHouse/blob/91d615812e809953142ad2854c768181e4a7b51c/src/Interpreters/MutationsInterpreter.cpp#L436
as its called by all others, what do you think ?

File: src/Interpreters/MutationsInterpreter.cpp [hunk 2]
```diff
@@ -1070,10 +1080,6 @@ void MutationsInterpreter::prepareMutationStages(std::vector<Stage> & prepared_s
 
     auto all_columns = storage_snapshot->getColumnsByNames(options, available_columns);
 
-    /// Add _row_exists column if it is present in the part
-    if (source.hasLightweightDeleteMask() || deleted_mask_updated)
```
Comment (SmitaRKulkarni) [inline src/Interpreters/MutationsInterpreter.cpp:1074]: Function `hasLightweightDeleteMask` not used anymore, can be removed

File: src/Interpreters/MutationsInterpreter.cpp (hunk 1 above)
Comment (SmitaRKulkarni) [reply src/Interpreters/MutationsInterpreter.cpp:423]: Looks like we are doing similar things in `getAvailableColumnsWithVirtuals`, maybe we can combine and use either this function or move everything to common constructor

**Review by SmitaRKulkarni — APPROVED** [review]
Rest all LGTM

File: src/Interpreters/MutationsInterpreter.cpp (hunk 1 above)
Comment (CurtizJ) [reply src/Interpreters/MutationsInterpreter.cpp:417]: This constructor with data part has a bit different logic. I'm afraid that I cannot unify them.

**Pull request merged**